### ISSUE TITLE
Add pathauto to core

### ISSUE THREAD

**Issue by maxpearl** [backdrop/backdrop-issues#87]
I know there have been a lot of arguments about this for D8, but in terms of the goals of backdrop, and ease of use, it seems that perhaps there is a good argument for this.

I'd love to see it happen. 

(And I don't know what the status of token is/will be...?)

**Comment by BWPanda** [comment]
I'd also like to suggest the addition of [Redirect](https://www.drupal.org/project/redirect)-type functionality as well (it's a recommended addition to Pathauto on D.o).

**Comment by enzolutions** [comment]
I did a version without drupal compatibility https://github.com/backdrop-contrib/pathauto @BWPanda you have documentation about that integration.

@quicksketch are you agree with integration with redirect, Sub Pathauto and Extended Path Aliases?

**Comment by quicksketch** [comment]
@enzolutions Sure, I think bundling them together would be fine. Long-term, I'd love to see all this functionality built-into path module directly, but we may need to prune some of the more difficult/edge-case code from the projects (e.g. bulk create and regenerate code) and move it into a separate contrib project.

**Comment by klonos** [comment]
Maybe we should keep the it simple in core and provide an "Advanced" section where we "hide" any "scary" settings from the main UI. I like how WP offers very basic settings through a simple UI:

![wp_permalink_settings](https://cloud.githubusercontent.com/assets/2423362/5900165/2177a6b6-a5ba-11e4-96a3-b27786ac1f54.png)

**Comment by quicksketch** [comment]
@jenlampton made a separate issue for implementing token into core at https://github.com/backdrop/backdrop-issues/issues/480.

Looking at the parts of Pathauto, I think it includes these main components:

- Path pattern administration.
- Hooks for various data types to create aliases (nodes, users, terms).
- Bulk creation/updates. 
- Bulk deletion.

The first two are obvious requirements for core and is what people generally think of as the functionality Pathauto provides. What about the last two for bulk create, update, and delete? Should that be included in core? I might be painting this based on my past negative experiences with bulk updates, but if that functionality is not 100% stable it may be better left to a separate contrib module.

**Comment by klonos** [comment]
> ...but if that functionality is not 100% stable it may be better left to a separate contrib module.

Perhaps that's exactly why it should be in core.

**Comment by quicksketch** [comment]
Touché. :wink:

**Comment by BWPanda** [comment]
I like and use the bulk update/delete functionality. One thing I'd suggest if we do add it to core (which I think we should) is that the Bulk Update functionality should include updating existing aliases (or at least have an option for this). It's annoying to have to bulk delete all aliases then bulk create them just because I updated something...

**Comment by BWPanda** [comment]
Also, we should add 'update' and 'delete' as VBO's to the alias page (admin/config/search/path) rather than as separate tabs (make consistent with the rest of Backdrop).

**Comment by klonos** [comment]
AFAICT, pathauto has the option to bulk update for some time now. There's no need to have any special option for it. If you want to update already existing aliases, all you need to do is select all relevant content from the Content admin page (```/admin/content```) and select the "Update URL alias" option available in the "Update options" drop-down menu.

Anything I'm missing here?

**Comment by klonos** [comment]
...or maybe you meant something like a "Save configuration and update existing aliases" button/option in the Patterns config page (```admin/config/search/path/patterns```). Oooh, I'd like that too! Yes please!!

**Comment by klonos** [comment]
...although, I don't know how this feature would work together with the option to not auto-generate aliases (https://www.drupal.org/project/pathauto_persist / [Merge in pathauto_persist module functionality to prevent losing manual aliases with node_save() calls](https://www.drupal.org/node/936222)). Another checkbox/option to respect the "Generate automatic URL alias" for each piece of content.

**Comment by BWPanda** [comment]
> Anything I'm missing here?

That only works for Content (nodes), but other entities can also have aliases that need updating (Taxonomy Terms, Users, etc.)

**Comment by jenlampton** [comment]
@klonos the bulk update action you are referring to is the one that @quicksketch mentioned as being "not 100% stable", and I think that's a polite way of putting it.  I've personally had very little success getting those bulk options to work - though I do wish for actually-working bulk actions at some point on pretty much every single site. Nobody wants to manually update paths for 10,000 blog posts!

I'd hate to see such a useful and in-high-demand feature be excluded. I'm not sure what the problem is with the current state of pathauto's bulk update operations, but I think we should aim to include that feature.  It's important.

**Comment by quicksketch** [comment]
> AFAICT, pathauto has the option to bulk update for some time now.

Ah, you're right. I must be thinking of a time in Drupal 5/6 era where the bulk creation tried to do updating also. It did not work well because I believe it tried to do all sorts of magic around whether a path should be "really updated" or not. I believe it also didn't integrate with Path Redirect, so you'd end up with 404s all over your site.

I'm fine with including the current options for bulk create and delete. It's the updating that's a real pain. I think that means that everything that's in Pathauto now is candidate to move into core.

>  I don't know how this feature would work together with the option to not auto-generate aliases (https://www.drupal.org/project/pathauto_persist / Merge in pathauto_persist module functionality to prevent losing manual aliases with node_save() calls).

I'd also like Pathauto persist to be part of core. I was campaigning/re-rolling in that issue years ago but I don't think it's going to happen.

**Comment by klonos** [comment]
> I was campaigning/re-rolling in that issue years ago but I don't think it's going to happen.

Yeah, I have noticed you in that issue. You can see me "subscribing" just a month before your first comment there :)

What do you mean you don't think it's happening? For D7 core you mean? ...or BD core?

**Comment by quicksketch** [comment]
> What do you mean you don't think it's happening?

Oh sorry that's not clear at all. I don't think it's going to happen in Drupal 7 contrib. That issue has been spinning tires for years. I'd definitely like to see this in the Backdrop port, and in core when Pathauto is merged into Path.

**Comment by serundeputy** [comment]
Here is a shot at a TODO list if these should be separate issues and they make sense I'm glad to do the leg work on making the issues and tackle some of them.  This is certainly not exhaustive just the first things I notice as I start inventory of the pieces of `pathauto`.
TODO:
- [ ] move `pathauto.settings.json` to `core/modules/path/config/path.settings.json`?
- [ ] should we combine pathauto.js and path.js?
- [ ] pathauto depends on token; strip out dependency or wait for token into core.
- [ ] Reomve README.md? Seems like the other core modules do not have READMEs.

**Comment by quicksketch** [comment]
Yep, all of those seem like the necessary and correct tasks @serundeputy.

I think Token and Pathauto are actually fairly loosely bound together. Core itself contains the `token_replace()` function. Token simply adds more token options and the UI for selecting a token. You could probably count on the token functions not even changing too much, so building this against the contrib Token module would probably allow this work to happen in parallel. We can adjust the (minor) differences once Token is fully in core.

**Comment by serundeputy** [comment]
@jenlampton @quicksketch 
I think this gets some of the grunt work out of the way:
https://github.com/serundeputy/backdrop/commit/82398324bf7a94aedf84bc25bf9d147748ee3e9d

It does not display the form to set the alias patterns though `admin/config/search/path/patterns` there are a number of places where I had questions about the right course of action.  I flagged those in the code with: `BACKDROP_MOVE_TO_CORE` at least half a dozen of those are namespacing of functions, this could be why the form does not display.

I hope this at least helps to get the ball rolling.

~Geoff

**Comment by jenlampton** [comment]
Thanks for getting started on this @serundeputy, I've got token mostly ported over at https://github.com/jenlampton/backdrop/tree/480-token-in-core and as soon as I have tests passing I'll come take a closer look at your notes :)

**Comment by serundeputy** [comment]
- [x] /admin/config/search/path/patterns form now displays for `Content Paths`, `Taxonomy Term Paths`, and `User Paths`.
  - [x] Form saves
- [x] node/add/% `URL path settings` Vertical Tab shows the `pathauto` version of the form
  - [ ] On `Save` does not apply the alias for `Generate automatic URL alias`
  - [x] If `Generate automatic URL alias` is unchecked normal alias is applied (fill out alias text box.)
- [x] /admin/structure/taxonomy/%/add `URL Path Settings` shows the `pathauto` version of the path form.
  - [ ] On `Save` does not apply the alias for `Generate automatic URL alias`
  - [x] If `Generate automatic URL alias` is unchecked normal alias is applied (fill out alias text box.)
- [ ] On `%/edit` forms the `pathauto` alias form is NOT displayed, the old `path` form is.
  - [ ] nodes
  - [ ] taxonomies

I swear I had the aliases updating and respecting the `pathauto` tokens at one point, but I can't seem to get back there :). Very `cold fusion` I know, but blarg.

**Comment by serundeputy** [comment]
update:
seems that on a fresh install of `backdrop` on branch [87/add-pathauto-to-core] (https://github.com/serundeputy/backdrop/tree/87/add-pathauto-to-core) the automatic aliases work for `basic page`, `article`, `taxonomy` terms, and `users`, but if you hit:
```
admin/config/search/path/patterns/settings
```
and change say separator to `-` and save then the automatic paths stop working?

so this branch `sometimes`™ works.

Any ideas welcome.

**Comment by quicksketch** [comment]
Thanks @serundeputy. Token has now been merged into core in https://github.com/backdrop/backdrop-issues/issues/480. I'm going to pick up your current branch and give it a spin.

**Comment by serundeputy** [comment]
:+1:

**Comment by quicksketch** [comment]
Lots of great progress on this today. I should be able to put up a pretty good PR for this by the weekend I hope. 

Great job @serundeputy on the port so far. You identified a bunch of problems with merging these two together, and the notes made it easy to see what you'd done when things got a little weird.

In my work, I'm trying to consolidate some of the weirder characteristics of Pathauto into a simpler API. As we can't maintain total compatibility (since the names of things are changing), it's a good opportunity to clean up a little.

- The old `hook_pathauto()` took an `$op` parameter, but there was only one supported op, "settings". Despite the name indicating it would generate a form (or so I thought), it actually returns an object of information about the token types. Given this hook's purpose, I renamed it to `hook_path_info()` and dropped the `$op` parameter.
- The `hook_pathauto_alias_types()` hook provided a 1-to-1 mapping of paths (e.g. `node/`, `user/`) to pathauto human labels (e.g `Content paths`, `User paths`). With the new generic purpose `hook_path_info()`, there's no need for these to be separate, so I consolidated this hook into `hook_path_info()`.
- There are a bunch of `$op` parameters all over the place. I think we should be able to eliminate all use of these. If we just pass in the entity objects directly, we can now check `$entity->isNew()`, so there's no need to inform the function if it's an "insert" or "update".

**Comment by pablo-romero** [comment]
If Path and Pathauto are going to be merged into one single module, will it include any kind of basic transliteration feature?

**Comment by quicksketch** [comment]
> If Path and Pathauto are going to be merged into one single module, will it include any kind of basic transliteration feature?

Not in the initial pass, but I think including transliteration module's functionality into core is also a great idea. Pathauto itself doesn't seem to include any direct transliteration support. However, these days, isn't including all kinds of characters in URLs entirely common? With entire domain names being in Chinese or Arabic, it seems like transliterating URLs might be something of the past. I'd love to hear your thoughts on it.

**Comment by pablo-romero** [comment]
I think that if it does not include it, maybe it could be good to display a note/warning, in case the site is not (or not only) in English.
I've had issues in the past when titles contain tildes and other characters like *ü, ó, è, ñ*, etc. that are quite common in Spanish, German, French, etc. Pathauto converts those characters by default into un-semantic strange ASCII strings (not good for SEO), that's why it's recommended to use the Transliteration module with any of those languages.

I think most users don't even know what transliteration means, and will probably expect Backdrop to do it for them out-of-the-box, e.g. like Wordpress does with slugs.

#884 can also open a new range of "strange" characters.

PS: Sorry for being a PITA with language-aware topics.

**Comment by klonos** [comment]
> ...it seems like transliterating URLs might be something of the past.

Not at all. Although non-latin domain names are allowed nowadays, some browsers support them better than others. So, it is a matter of browser choice and although this fact makes this a non-issue (80/20 case argument if people use the most popular instead of some "exotic" ones), you need to consider that URLs are not "consumed" only by browsing applications.

Then again, risking you to say that this is FUD, some applications and people treat non-latin URLs as spam.

Related: [Are Foreign Language Characters in URLs Bad for SEO?](https://www.fusionbox.com/blog/detail/are-foreign-language-characters-in-urls-bad-for-seo/536/)

Lastly, it is also a matter of preference. Some people for example do not like switching their keyboard layout while typing the URL (in order to type the domain in latin and some part of the rest of the URL in non-latin). They simply find it tedious.

**Comment by quicksketch** [comment]
Thanks @pablo-romero and @klonos for that great input. I've filed a followup for including Transliteration in core (https://github.com/backdrop/backdrop-issues/issues/904). Unfortunately it looks like this task is fairly involved, so we'll have to do it post-1.1.0.

As far as Pathauto in core goes, I've taken @serundeputy's initial work and consolidated a bunch of functionality into the entity system. This has made it so that the following hooks are all consolidated:

- hook_taxonomy_term_insert/hook_node_insert/hook_user_insert => hook_entity_insert
- hook_taxonomy_term_update/hook_node_update/hook_user_update => hook_entity_update
- hook_taxonomy_term_delete/hook_node_delete/hook_user_delete => hook_entity_delete

To further consolidation, I made a new helper function `path_form_element()` that can add the Pathauto fieldset to any existing element form, including Taxonomy and Nodes out of the box (to match D7).

And lastly to help the consistency of Pathauto, I built-in [Pathauto Persist](https://www.drupal.org/project/pathauto_persist) directly as part of the core path system by adding an "auto" column directly to the `path_alias` table. Now all entities will properly remember if they have auto-generated aliases or manually defined ones, preventing the [very common problem](https://www.drupal.org/node/936222) with Pathauto. 

The PR at https://github.com/backdrop/backdrop/pull/869 is functionally complete. There are a few lingering test failures that I should be able to clear up shortly. @serundeputy, if you have the chance, I'd love to have a review of the PR code and try it out locally.

**Comment by klonos** [comment]
> And lastly to help the consistency of Pathauto, I built-in Pathauto Persist directly as part of the core path system...

:tada:

**Comment by docwilmot** [comment]
Are these new hooks replacing the originals or in addition to the originals? I'm thinking change records.

**Comment by quicksketch** [comment]
> Are these new hooks replacing the originals or in addition to the originals? I'm thinking change records.

The changes to use hook_entity_[action]() are all within path.module and shouldn't affect the API of any other module. The changes up above in https://github.com/backdrop/backdrop-issues/issues/87#issuecomment-97716995 are replacements for the old `hook_pathauto_*` hooks.

**Comment by serundeputy** [comment]
installed beautifully on a fresh backdrop 1.x;
update.php ran great!

  aliases for 
    * Article
    * Basic Pages
    * Users
    * Taxonomy Terms

I can manually set aliases (uncheck the box), and reset to auto (check the box) all working perfectly.

working as expected.  Respecting patterns and tokens from `/admin/config/search/path/patterns`.
Respecting character separator.  

Spun up a few hundred Articles, Basic Pages, Taxonomy Terms and Users w/ `devel` all taking aliases as expected!

A few `php` warnings on form saves:
![screen shot 2015-05-07 at 7 28 58 pm](https://cloud.githubusercontent.com/assets/6763671/7528265/7633834a-f4f3-11e4-9548-b73a8f503219.png)
![screen shot 2015-05-07 at 7 39 10 pm](https://cloud.githubusercontent.com/assets/6763671/7528266/76340a22-f4f3-11e4-8fce-5f691d8ea407.png)

**Comment by quicksketch** [comment]
Great, thanks @serundeputy! It looks like that particular issue is a problem with the token integration, not this set of Pathauto changes. I filed a new issue at https://github.com/backdrop/backdrop-issues/issues/906.

**Comment by serundeputy** [comment]
with pathauto contrib and then `git co 87/pathauto`:
![screen shot 2015-05-07 at 9 19 58 pm](https://cloud.githubusercontent.com/assets/6763671/7528966/07de8884-f4ff-11e4-9537-a8ed65f65429.png)

**Comment by serundeputy** [comment]
I think this could be a `settings.php` and `files/conf_` more than `pathauto`, `token` upgrade problem.

I'll try to detangle tomorrow.

**Comment by serundeputy** [comment]
@quicksketch seems i'm still getting the error in the screenshot: can not redeclare `token_type_load()`.

here is what i did.
 * installed backdrop-1.0.5
 * installed token
 * installed pathauto
 * made a page ; checked to see that it had automatic alias

pulled down a copy of backdrop 1.x in a different directory;
  * moved the core directory out of my backdrop-1.0.5 installation
  * replaced with the core directory of the 1.x branch;
  * visit site in browser.

~Geoff

**Comment by quicksketch** [comment]
Hi @serundeputy, I made a few fixes in the 1.x branch that might help prevent WSOD on the update.php page, including #910 and #722. I tried this all from scratch again today:

- Fresh install of 1.0.5. Install Pathauto and Token latest releases.
- Configure Pathauto to add aliases for different types of content. Adjust any settings as well.
- Git checkout 87/pathauto to replace 1.0.5.
- Visit the homepage of the site. At this point I get this warning:
![token_max_depth](https://cloud.githubusercontent.com/assets/100206/7546981/0b78c6aa-f5a2-11e4-8f78-fefad26daf16.png)
  Some other pages will also cause errors, such as attempting to save a node or visiting the path pattern settings pages. These are usually fatal duplicate function errors.
- However, at this point I can also visit update.php (with no errors if I do this first). And running update.php works successfully with no errors. Visiting path pattern settings I can see that all aliases are maintained and work correctly.

**Comment by quicksketch** [comment]
Hi @serundeputy, I was writing the above post when you posted as well. I'm not sure what to do here. It doesn't seem possible that we could totally prevent duplicate function errors between token.module and token.inc unless we rename the functions (thus causing more compatibility breaks). Historically, errors on your site prior to running update.php have been considered "normal", so if we can at least get users to update.php without errors than that's the minimum level of acceptability.

Of course, we can (and probably will) also recommend that any users have Token or Pathauto modules remove those modules from their installation prior to doing the update to 1.1.0.

**Comment by quicksketch** [comment]
Ah! I got it. We can add the list of merged modules into `module_list()`, specifically removing ones that have been merged into core. That will make it so that Backdrop does not load the .module file for these modules. I'll update the PR and then we can give this another go.

**Comment by quicksketch** [comment]
@serundeputy This latest commit should prevent the old Token or Pathauto modules from be executed, even if they are still enabled and present:
https://github.com/quicksketch/backdrop/commit/42811dbe6f810efa2b89d4d9aa05bc6912ff7306

In my testing, this prevents all critical errors on the site before update.php is executed. It should correct the problems with update.php itself as well, as the duplicate functions should all be avoided.

**Comment by klonos** [comment]
I manually patched the files from commit https://github.com/quicksketch/backdrop/commit/42811dbe6f810efa2b89d4d9aa05bc6912ff7306 because I was getting the same ```Cannot redeclare token_type_load()``` fatal error as @serundeputy 

The site loads fine, but trying to visit the site status page (/admin/reports/status) throws another fatal:

```
Fatal error: Cannot redeclare token_get_token_problems() (previously declared in C:\www\backdrop\core\includes\token.inc:404) in C:\www\backdrop\modules\token\token.install on line 332
Call Stack
#   Time    Memory   Function                                      Location
1   0.0002  244200   {main}( )                                                                          ...\index.php:0
2   0.7478  13877040 menu_execute_active_handler( )                                                     ...\index.php:21
3   0.7479  13878192 layout_route_handler( )                                                            ...\menu.inc:520
4   0.8059  14453304 LayoutRendererStandard->render( )                                                  ...\layout.module:445
5   0.8059  14455856 LayoutRendererStandard->renderLayout( )                                            ...\layout_renderer_standard.inc:312
6   0.8061  14460968 LayoutRendererStandard->renderBlocks( )                                            ...\layout_renderer_standard.inc:331
7   0.8061  14461920 LayoutRendererStandard->renderBlock( )                                             ...\layout_renderer_standard.inc:408
8   0.8061  14462000 BlockLegacy->getContent( )                                                         ...\layout_renderer_standard.inc:450
9   0.8061  14462264 BlockLegacy->buildBlock( )                                                         ...\block_legacy.class.inc:57
10  0.8061  14463152 module_invoke( )                                                                   ...\block_legacy.class.inc:30
11  0.8061  14464248 call_user_func_array:{C:\www\backdrop\core\includes\module.inc:859} ( )            ...\module.inc:859
12  0.8061  14464832 system_block_view( )                                                               ...\module.inc:859
13  0.8103  14941344 call_user_func_array:{C:\www\backdrop\core\modules\system\system.module:2231} ( )  ...\system.module:2231
14  0.8104  14941664 system_status( )                                                                   ...\system.module:2231
15  0.8143  15152112 backdrop_load_updates( )                                                           ...\system.admin.inc:2166
16  0.8301  16518608 module_load_install( )                                                             ...\install.inc:84
17  0.8301  16518656 module_load_include( )                                                             ...\module.inc:296
```

**Comment by klonos** [comment]
...also tried downloading the whole tree, but same thing and another fatal when trying the "Run updates" page (/core/update.php?op=info):

```
Fatal error: Cannot redeclare token_get_token_problems() (previously declared in C:\www\backdrop\core\includes\token.inc:404) in C:\www\backdrop\modules\token\token.install on line 332
Call Stack
#  Time    Memory    Function                  Location
1  0.0011  329400    {main}( )                 ...\update.php:0
2  0.4213  13924040  backdrop_load_updates( )  ...\update.php:484
3  0.4404  15242576  module_load_install( )    ...\install.inc:84
4  0.4404  15242624  module_load_include( )    ...\module.inc:296
```

**Comment by quicksketch** [comment]
Thanks @klonos! I hadn't thought that we'd need to do the same thing to prevent .install files from being loaded, but now that you mention the problems it seems obvious that this would happen. I'll work on finding a way to prevent token.install from loading as well.

**Comment by quicksketch** [comment]
I've updated the PR to include a fix to prevent .install files for "merged" modules from being loaded in update.php or the status report page. I amended the last commit, so if you're applying things manually the only thing that changed was the addition to `backdrop_load_update()`: https://github.com/quicksketch/backdrop/commit/118e9006c930c7c63d890af8c33d1d83bae4e007#diff-eb90f57f441b14b60d6f4670ba018affL80

**Comment by serundeputy** [comment]
this works beautifully ; thanks so much @quicksketch 
  [ ] upgraded from 1.0.5 to 87/pathauto w/ no contrib modules
  [-] upgraded from 1.0.5 to 87/pathauto w/ token installed
  [-] upgraded from  1.0.5 to 87/pathauto w/ token and pathauto installed
    * preserved existing paths

:smile: :bow: :smile:

**Comment by quicksketch** [comment]
Woot woot! I'll probably merge this in later this evening. Everything else we can get in before Friday will just be icing on the :cake:.

**Comment by klonos** [comment]
Manually patched diff and now the "Status report" and "Run updates" pages are accessible once again. Thanx!

PS: While running the update.php, the section of the pending updates reads:

>**path module**
- 1000 - Upgrades Drupal 7 Pathauto variables to Path config.
- 1001 - Moves Backdrop Pathauto configuration to Path module configuration.

Is the "Drupal 7" intentional there? Asking because my test site was not a D7 upgrade - simply a Backdrop 1.0.0 that I upgrade over time as new releases/patches come out.

PPS: #913

**Comment by quicksketch** [comment]
> Is the "Drupal 7" intentional there? Asking because my test site was not a D7 upgrade - simply a Backdrop 1.0.0 that I upgrade over time as new releases/patches come out.

Yes, "Drupal 7" is intentional. We provide an upgrade path directly from Drupal 7 AND Backdrop 1.0.0. So if regardless of whether you used Backdrop's Pathauto or D7's Pathauto, both are migrated.

**Comment by quicksketch** [comment]
From the looks of the bug reports @klonos and others have been filing (here and in other issues), most of the problems introduced are due to Token's merging (because of either function overlaps or misported tokens). Pathauto itself so far seems to be working well. To get us "feature complete" and able to test everything together, I've merged this into 1.x. woot woot!

**Comment by klonos** [comment]
Thanx, that's great!
